druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,218 @@ use crate::kurbo::Point;
     ///
-    /// This is in px units, that is, adjusted for hi-dpi.
+    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
     pub pos: Point,
-    /// Keyboard modifiers at the time of the mouse event.
+    /// Mouse buttons being held down during a move or after a click event.
+    /// Thus it will contain the `button` that triggered a mouse-down event,
+    /// and it will not contain the `button` that triggered a mouse-up event.
+    pub buttons: MouseButtons,
+    /// Keyboard modifiers at the time of the event.
     pub mods: KeyModifiers,
     /// The number of mouse clicks associated with this event. This will always
-    /// be `0` for a mouse-up event.
-    pub count: u32,
-    /// The currently pressed button in the case of a move or click event,
-    /// or the released button in the case of a mouse-up event.
+    /// be `0` for a mouse-up and mouse-move events.
+    pub count: u8,
+    /// The button that was pressed down in the case of mouse-down,
+    /// or the button that was released in the case of mouse-up.
+    /// This will always be `MouseButton::None` in the case of mouse-move.
     pub button: MouseButton,
 }
 
 /// An indicator of which mouse button was pressed.
 #[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[repr(u8)]
 pub enum MouseButton {
     /// Left mouse button.
     Left,
-    /// Middle mouse button.
-    Middle,
     /// Right mouse button.
     Right,
+    /// Middle mouse button.
+    Middle,
     /// First X button.
     X1,
     /// Second X button.
     X2,
+    /// No mouse button.
+    None,
 }
 
 impl MouseButton {
-    /// Returns `true` if this is the left mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Left`].
+    ///
+    /// [`MouseButton::Left`]: #variant.Left
+    #[inline]
     pub fn is_left(self) -> bool {
         self == MouseButton::Left
     }
 
-    /// Returns `true` if this is the right mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Right`].
+    ///
+    /// [`MouseButton::Right`]: #variant.Right
+    #[inline]
     pub fn is_right(self) -> bool {
         self == MouseButton::Right
     }
+
+    /// Returns `true` if this is [`MouseButton::Middle`].
+    ///
+    /// [`MouseButton::Middle`]: #variant.Middle
+    #[inline]
+    pub fn is_middle(self) -> bool {
+        self == MouseButton::Middle
+    }
+
+    /// Returns `true` if this is [`MouseButton::X1`].
+    ///
+    /// [`MouseButton::X1`]: #variant.X1
+    #[inline]
+    pub fn is_x1(self) -> bool {
+        self == MouseButton::X1
+    }
+
+    /// Returns `true` if this is [`MouseButton::X2`].
+    ///
+    /// [`MouseButton::X2`]: #variant.X2
+    #[inline]
+    pub fn is_x2(self) -> bool {
+        self == MouseButton::X2
+    }
+}
+
+/// A set of [`MouseButton`]s.
+///
+/// [`MouseButton`]: enum.MouseButton.html
+#[derive(PartialEq, Eq, Clone, Copy, Default)]
+pub struct MouseButtons(u8);
+
+impl MouseButtons {
```

Comment:
just sort of riffing on this type; if the idea is to construct it from `MouseButton`s, one possible thing to do is implement `FromIterator<MouseButton>`, and then you could write, say,

```rust
let buttons: MouseButtons = [MouseButton::Left, MouseButton::X1].iter().copied().collect();
```

That doesn't feel great really; probably simpler to have a constructor that just takes a `&[MouseButton]`.

In any case I'm just daydreaming, this isn't necessary or anything.

Reply:
The following one liner isn't too bad either:
```rust
let buttons = MouseButtons::new().with(MouseButton::Left).with(MouseButton::Right);
```

There can be value in adding more methods, however I think I would leave these slice/iterator  constructors out of this PR.